### ISSUE TITLE
Windows Builds

### ISSUE THREAD

**Issue by jakelandis** [elastic/logstash#7767]
Windows builds were recently introduced to the Logstash Jenkins builds, and have not fully worked. 

For example: https://logstash-ci.elastic.co/job/elastic+logstash+master+multijob-windows-compatibility/

Below is the list to get Windows builds fully functional:
- [x] Fix the Java core tests https://github.com/elastic/logstash/pull/7716
- [x] Fix Ingest Conversion tool tests https://github.com/elastic/logstash/issues/7642
- [x] Fix Jenkins ```$HOME``` to ensure proper escaping. https://github.com/elastic/infra/pull/2483
- [x] Allow long file names on Jenkins. https://github.com/elastic/infra/issues/2487
(https://github.com/elastic/logstash/issues/7634#issuecomment-316843373)
- [x] Fix Null Pointer from ```dsl.instance_eval``` issue. See [this comment](https://github.com/elastic/logstash/issues/7634#issuecomment-316895648), and https://github.com/elastic/logstash/issues/7768
- [x] [6.0, master] Fix Ruby test failures #8116 #8074
- [x] [5.6] Fix Ruby test failures #8125 
- [x] [5.5] Fix build & tests #8354 
- [x] fix `logstash.bat`, `setup.bat` and `rspec.bat` #8126  #8178 #8359
- [x] Evaluate if builds need to be run on Windows 2012. See [this comment](https://github.com/elastic/infra/issues/2487#issuecomment-317078580) and the conversation below.
- [x] Fix false positive ```success``` from ci\unit_tests.bat See [this comment](https://github.com/elastic/logstash/issues/7634#issuecomment-316843373) and [this comment](https://github.com/elastic/logstash/issues/7767#issuecomment-333928201)
- [x] Support builds on Windows 2016 w/ longfile name fix. 
- [x] Support builds on Windows 2012 (no long file name fix available) #8398 
- [x] Discuss updating the [support matrix](https://www.elastic.co/support/matrix) for Windows to include a footnote about the long file names and potential workarounds. (see below - logged follow up https://github.com/elastic/logstash/issues/8424)

-----------
My notes to run locally:

 ensure that JRuby is installed at C:
 For example: C:\jruby-9.1.12.0
 Need to run from cmd, not powershell, but can launch cmd from powershell for a better cmd.
 From powershell:
 [clone the repo](https://www.develves.net/blogs/asd/articles/using-git-with-powershell-on-windows-10)

    cmd
    set JRUBYSRCDIR=C:\
    ci\ci_test.bat

**Comment by suyograo** [comment]
@jakelandis we need to support Windows 2012. See our support matrix here: https://www.elastic.co/support/matrix

**Comment by jakelandis** [comment]
Is there a distinction between supporting Windows 2012 and running CI builds on Windows 2012 ?

**Comment by suyograo** [comment]
@jakelandis Ideally, we'd be running our tests on all the platforms we support.

**Comment by jakelandis** [comment]
@suyograo - agreed, ideally. However,  If we want to run on Windows 2012 we will need to work with infra to either shorten the base names of the builds, or provide consistent substitutions that we can leverage in our .bat file. 

For example:
```
subst m: C:\Users\jenkins\workspace\elastic+logstash+pull-request+multijob-windows-compatibility
subst n: C:\Users\jenkins\workspace\elastic+logstash+master+multijob-windows-compatibility
```
Then in our .bat file we would have to check the current directory (```%~dp0\..```) and based on some static mapping change to the approriate drive (e.g. m:) before running our tests.  I played around with doing this on the fly, but there are potential issues of collision without a static set of mapping integrated into OS as startup time. 

tl;dr -Builds on Windows 2012 is possible, we just need to work for it.  I will add it to the list above.

**Comment by dliappis** [comment]
FYI a new Windows 2012_r2 slave has been launched, see details in this infra comment: https://github.com/elastic/infra/issues/2487#issuecomment-318323916

**Comment by colinsurprenant** [comment]
#7768 fixed by #8068

**Comment by colinsurprenant** [comment]
Windows build fixes in #8116

**Comment by colinsurprenant** [comment]
Windows build fixes for 5.6 #8125

**Comment by colinsurprenant** [comment]
Launch script rework #8178

**Comment by colinsurprenant** [comment]
If we want to move forward we'll need reviewers for #8125 and #8178.

**Comment by jakelandis** [comment]
We now have the Windows builds fully integrated into the Jenkins builds (5.5+) .  The Java unit tests work without issues. The Ruby unit tests are only functional on Windows 2016 with the long file name fixes. 

The Ruby unit tests that are running on Windows 2012 are failing in the setup, but still show as green due to the false positive success issue. 

..so close.  The outstanding items are to find a workaround to the long file name issue and fix the false positive issue (in that order, else we turn our builds red again).

**Comment by jakelandis** [comment]
Followup issue logged : https://github.com/elastic/logstash/issues/8383 (Add Windows to set of Acceptance tests)

**Comment by colinsurprenant** [comment]
PR #8398 in to attempt to fix the long path issues.  Once this is solved I will do the false positives.

**Comment by colinsurprenant** [comment]
So #8398 is merged and solves the long path issue and should also solve the false positive with the added checks for rake tasks returned `errorlevel` - I tested locally for test failures and these correctly resulted in
```
Error: failed to run core tests. Aborting..
```
which also means exiting the shell with `exit /B 1` which should be picked up by Jenkins as an error condition.

**Comment by colinsurprenant** [comment]
@jakelandis for the last task 
> Update the support matrix for Windows to include a footnote about the long file names and potential workarounds.

What did you have in mind? do you mean to tell users that using logstash in a too deep path could result in errors and should use a `subst` drive?

**Comment by jakelandis** [comment]
> What did you have in mind?

@colinsurprenant - Not sure, I think I just wanted to have the conversation if it was needed. 

@dedemorton - what are your thoughts if we should add an asterisk to the Windows 2012/2016 compatibility matrix ?

Windows has a 260 character limit for files and paths. Logstash uses some very long paths  (see below).  Windows 2016 has a registry settings that can be set to get around this limitation, and Window 2012 there is a different, but clunky workaround .  The Windows limit is well known, but it may not be well known that Logstash uses very long paths. 

I ran [this script](http://blog.danskingdom.com/powershell-script-to-check-path-lengths) to see exactly where we are at.

As you can see below the the longest file is 204 characters in length, meaning that the the root can be 56 characters long. If installed to "C:\Program Files\elastic\logstash" we would still be under the limit.

I _think_ we are good without an asterisk, since for the common use-case we are still under the limit. But will defer to @dedemorton or others opinions.

<img width="1586" alt="image" src="https://user-images.githubusercontent.com/976291/31037261-b60ff102-a535-11e7-85ca-db70349b53c2.png">

**Comment by dedemorton** [comment]
@jakelandis 

I think we are good without mentioning the limitation in the matrix if we:

* Mention the limitation somewhere in the installation docs.
* Document the workaround(s) for those poor unfortunate users who do run into issues related to long paths.
* Provide a meaningful error message so users understand why they are having problems. Is this possible?

**Comment by jakelandis** [comment]
> Provide a meaningful error message so users understand why they are having problems. Is this possible?

I don't think this is possible with a decent amount of investment of time. 


> Mention the limitation somewhere in the installation docs.

and

> Document the workaround(s) for those poor unfortunate users who do run into issues related to long paths.

Any suggestions on where this should go ? (I can provide the content)

**Comment by dedemorton** [comment]
We can add a section a Windows section to the following topic:

**Comment by dedemorton** [comment]
@jakelandis 

> Any suggestions on where this should go ? (I can provide the content)

We could add a section called "Installing on Windows" to this topic: https://www.elastic.co/guide/en/logstash/current/installing-logstash.html

Since we don't have a troubleshooting section in the Logstash docs (yet), we can try adding the the troubleshooting info to the Windows section and see if it works there...really depends on how long it is.

**Comment by jakelandis** [comment]
@dedemorton - +1 to adding this to the installation guide.  I logged https://github.com/elastic/logstash/issues/8424 to address this.

**Comment by jakelandis** [comment]
@colinsurprenant - I don't think the false positive is fixed yet. I found a case where the Ruby unit tests failed, but the job still ended in success: https://logstash-ci.elastic.co/job/elastic+logstash+6.0+multijob-windows-compatibility/112/console

**Comment by colinsurprenant** [comment]
@jakelandis damn 😩 - looking into it.

**Comment by colinsurprenant** [comment]
@jakelandis confirmed - the problem is with `rspec.bat` which does not propagate the failed exit status. Fixed in #8450

**Comment by colinsurprenant** [comment]
I think we're all good to close now! Thanks all for your help!
